lib/services/github.go

```diff
@@ -18,15 +18,110 @@ package services
+
+// githubConnectorMutex is a mutex for the Github auth connector
+// registration functions.
+var githubConnectorMutex sync.RWMutex
+
+// GithubAuthCreator creates a new Github connector.
+type GithubAuthCreator func(string, types.GithubConnectorSpecV3) (types.GithubConnector, error)
+
+var githubAuthCreator GithubAuthCreator
+
+// RegisterGithubAuthCreator registers a function to create Github auth connectors.
+func RegisterGithubAuthCreator(creator GithubAuthCreator) {
+	githubConnectorMutex.Lock()
+	defer githubConnectorMutex.Unlock()
+	githubAuthCreator = creator
+}
+
+// NewGithubConnector creates a new Github auth connector.
+func NewGithubConnector(name string, spec types.GithubConnectorSpecV3) (types.GithubConnector, error) {
+	githubConnectorMutex.RLock()
+	defer githubConnectorMutex.RUnlock()
+	return githubAuthCreator(name, spec)
+}
+
+// GithubAuthInitializer initializes a Github auth connector.
+type GithubAuthInitializer func(types.GithubConnector) (types.GithubConnector, error)
+
+var githubAuthInitializer GithubAuthInitializer
+
+// RegisterGithubAuthCreator registers a function to initialize Github auth connectors.
+func RegisterGithubAuthInitializer(init GithubAuthInitializer) {
+	githubConnectorMutex.Lock()
+	defer githubConnectorMutex.Unlock()
+	githubAuthInitializer = init
+}
+
+// InitGithubConnector initializes c and returns a [types.GithubConnector]
+// ready for use. InitGithubConnector must be used to initialize any
+// uninitialized [types.GithubConnector]s before they can be used.
+func InitGithubConnector(c types.GithubConnector) (types.GithubConnector, error) {
+	githubConnectorMutex.RLock()
+	defer githubConnectorMutex.RUnlock()
+	return githubAuthInitializer(c)
+}
+
+// GithubAuthConverter converts a Github auth connector so it can be
+// sent over gRPC.
+type GithubAuthConverter func(types.GithubConnector) (*types.GithubConnectorV3, error)
+
+var githubAuthConverter GithubAuthConverter
+
+// RegisterGithubAuthCreator registers a function to convert Github auth connectors.
+func RegisterGithubAuthConverter(convert GithubAuthConverter) {
+	githubConnectorMutex.Lock()
+	defer githubConnectorMutex.Unlock()
+	githubAuthConverter = convert
+}
+
+// GithubAuthConverter converts a Github auth connector so it can be
+// sent over gRPC.
+func ConvertGithubConnector(c types.GithubConnector) (*types.GithubConnectorV3, error) {
+	githubConnectorMutex.RLock()
+	defer githubConnectorMutex.RUnlock()
+	return githubAuthConverter(c)
+}
+
+func init() {
+	RegisterGithubAuthCreator(types.NewGithubConnector)
+	RegisterGithubAuthInitializer(func(c types.GithubConnector) (types.GithubConnector, error) {
+		return c, nil
+	})
+	RegisterGithubAuthConverter(func(c types.GithubConnector) (*types.GithubConnectorV3, error) {
+		connector, ok := c.(*types.GithubConnectorV3)
+		if !ok {
+			return nil, trace.BadParameter("unrecognized github connector version %T", c)
+		}
+		return connector, nil
+	})
+}
+
```

Comment:
Broadly speaking I'm finding the design of this section confusing. What I think is going on is that you're relying on this `init()` function to register these global `var`iables wherever this package is first imported, and then calling these exported functions which in turn make use of the global variables. AFAICT this will work because everything is mutex-protected. 

But its a very awkward design because it leaves these global variables just sort of swimming ephemerally in memory, without any explicit relation to the other memory objects that depend on them. As a (rough) analogy, imagine a blueprint for a car which doesn't specify that the wheels will be attached to the axles.

I was going to make some specific suggestions about how to change this, but now that I'm looking more closely, frankly I'm stumped about what you're ultimately aiming at here. Is the idea to eventually be able to call these `RegisterGithubAuth*` functions from somewhere else other than this `init()`, in order to swap them out on the fly while the program is running?

Reply:
@capnspacehook my mistake, I just found that there's a corresponding PR in enterprise https://github.com/gravitational/teleport.e/pull/527 where you use the `RegisterGithubAuth*` functions, and that you're copying a design pattern that we already have in place. I remain a bit skeptical of that overall approach, but I see how it works and its not really yours to defend.

Reply:
Sorry for the confusion, will do